p2p/libp2p/networksharding/listKadSharder.go

```diff
@@ -0,0 +1,164 @@
+package networksharding
+
+import (
+	"fmt"
+	"math/big"
+	"sort"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/p2p"
+	"github.com/ElrondNetwork/elrond-go/sharding"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+const minAllowedConnectedPeers = 2
+const minAllowedPeersOnList = 1
+
+// listKadSharder is the struct able to compute an eviction list of connected peers id according to the
+// provided parameters. It basically splits all connected peers into 3 lists: intra shard peers, cross shard peers
+// and unknown peers by the following rule: both intra shard and cross shard lists are upper bounded to provided
+// maximum levels, unknown list is able to fill the gap until maximum peer count value is fulfilled.
+type listKadSharder struct {
+	peerShardResolver p2p.PeerShardResolver
+	crtPeerId         peer.ID
+	maxPeerCount      int
+	maxIntraShard     int
+	maxCrossShard     int
+}
+
+// NewListKadSharder creates a new kad list based kad sharder instance
+func NewListKadSharder(
+	resolver p2p.PeerShardResolver,
+	crtPeerId peer.ID,
+	maxPeerCount int,
+	maxIntraShard int,
+	maxCrossShard int,
+) (*listKadSharder, error) {
+	if resolver == nil {
+		return nil, p2p.ErrNilPeerShardResolver
+	}
+	if maxPeerCount < minAllowedConnectedPeers {
+		return nil, fmt.Errorf("%w, maxPeerCount should be at least %d", p2p.ErrInvalidValue, minAllowedConnectedPeers)
+	}
+	if maxIntraShard < minAllowedPeersOnList {
+		return nil, fmt.Errorf("%w, maxIntraShard should be at least %d", p2p.ErrInvalidValue, minAllowedPeersOnList)
+	}
+	if maxCrossShard < minAllowedPeersOnList {
+		return nil, fmt.Errorf("%w, maxCrossShard should be at least %d", p2p.ErrInvalidValue, minAllowedPeersOnList)
+	}
+
+	return &listKadSharder{
+		peerShardResolver: resolver,
+		crtPeerId:         crtPeerId,
+		maxPeerCount:      maxPeerCount,
+		maxIntraShard:     maxIntraShard,
+		maxCrossShard:     maxCrossShard,
+	}, nil
+}
+
+// ComputeEvictList returns the eviction list
+func (lks *listKadSharder) ComputeEvictList(pid peer.ID, connected []peer.ID) []peer.ID {
+	allPeers := connected
+	if !lks.Has(pid, connected) {
+		allPeers = append(connected, pid)
+	}
+
+	evictionProposed := make([]peer.ID, 0)
+	intraShard, crossShard, unknownShard := lks.splitPeerIds(allPeers)
+
+	intraShard, e := lks.evict(intraShard, lks.maxIntraShard)
+	evictionProposed = append(evictionProposed, e...)
+
+	crossShard, e = lks.evict(crossShard, lks.maxCrossShard)
+	evictionProposed = append(evictionProposed, e...)
+
+	sum := len(intraShard) + len(crossShard) + len(unknownShard)
+	if sum <= lks.maxPeerCount {
+		return evictionProposed
+	}
+	_, e = lks.evict(unknownShard, lks.maxPeerCount+1-len(intraShard)-len(crossShard))
```

Comment:
do the computation before and add to a variable.

Reply:
done